File: Hihocoder/1305.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <queue>
#include <map>
#include <unordered_set>
#include <unordered_map>
#include <cstring>
#include <bitset>
#include <algorithm>

using namespace std;

const int N = 100100;

struct Interval {
	int start, end;
	Interval(): start(0), end(0) {}
	Interval(int s, int e): start(s), end(e) {}
	bool operator < (const Interval& interval) const {
        return (start < interval.start);
    }
} A[N], B[N];

int merge(Interval* interval, int n) {
	if (n <= 0) {
		return 0;
	}

	int idx = 0;
	sort(interval, interval + n);

	Interval current = interval[0];
	for (int i = 0; i < n; i++) {
		if (interval[i].start > current.end) {
			interval[idx++] = current;
			current = interval[i];
		} else {
			current.end = max(current.end, interval[i].end);
		}
	}
	interval[idx++] = current;
	return idx;
}

int diff(Interval* A, Interval* B, int n, int m) {
	if (n <= 0 || m <= 0) {
		return 0;
	}

	int ret = 0, i = 0, j = 0;
	while (i < n && j < m) {
		if (A[i].end <= B[j].start) {
			ret += A[i].end - A[i].start;
			i++;
		} else if (A[i].start >= B[j].end) {
			j++;
		} else if (B[j].end >= A[i].end) {
			int len = A[i].end - B[j].start;
			ret += max(0, A[i].end - A[i].start - len);
			i++;
		} else {
			int len = B[j].end - B[j].start;
			ret += max(0, B[j].end - A[i].start - len);
			A[i].start = B[j].end;
			j++;
		}
	}
	while (i < n) {
		ret += A[i].end - A[i].start;
		i++;
	}
	return ret;
}

int main() {
	// freopen("input", "r", stdin);
	int n, m; scanf("%d%d", &n, &m);

	for (int i = 0; i < n; i++) {
		int start, end; scanf("%d%d", &start, &end);
		A[i].start = start, A[i].end = end;
	}

	for (int i = 0; i < m; i++) {
		int start, end; scanf("%d%d", &start, &end);
		B[i].start = start, B[i].end = end;
	}

	n = merge(A, n);
	m = merge(B, m);
	printf("%d\n", diff(A, B, n, m));
	return 0;
}

File: Hihocoder/1305.java
import java.io.*;
import java.math.BigInteger;
import java.util.*;

public class Main {

    public static ArrayList<Interval> merge(Interval[] interval) {
        ArrayList<Interval> ret = new ArrayList<Interval>();
        int n = interval.length;
        int start = interval[0].start, end = interval[0].end;
        for(int i = 1; i < n; i++) {
            if(interval[i].start > end) {
                ret.add(new Interval(start, end));
                start = interval[i].start;
                end = interval[i].end;
            }
            else {
                end = Math.max(interval[i].end, end);
            }
        }
        ret.add(new Interval(start, end));
        return ret;
    }

    public static void main(String[] args) throws FileNotFoundException {
//        Scanner input = new Scanner(new File("test.txt"));
        Scanner input = new Scanner(System.in);
        int n = input.nextInt(), m = input.nextInt();
        Interval[] original = new Interval[n];
        Interval[] delete = new Interval[m];
        for(int i = 0; i < n; i++)
            original[i] = new Interval(input.nextInt(), input.nextInt());
        for(int i = 0; i < m; i++)
            delete[i] = new Interval(input.nextInt(), input.nextInt());
        Comparator<Interval> compare = new Comparator<Interval>(){
            public int compare(Interval i1, Interval i2) {
                if(i1.start == i2.start)
                    return i1.end - i2.end;
                else
                    return i1.start - i2.start;
            }
        };
        Arrays.sort(original, compare);
        Arrays.sort(delete, compare);

        ArrayList<Interval> inter1 = merge(original);
        ArrayList<Interval> inter2 = merge(delete);

        n = inter1.size();
        m = inter2.size();
        int i = 0, j = 0;
        int total = 0, minus = 0;
        while(i < n && j < m) {
            if(inter1.get(i).start >= inter2.get(j).end) {
                j++;
            }
            else if(inter1.get(i).end <= inter2.get(j).start) {
                total += inter1.get(i).end - inter1.get(i).start;
                i++;
            }
            else {
                minus += Math.min(inter1.get(i).end,  inter2.get(j).end) - Math.max(inter1.get(i).start, inter2.get(j).start);
                if(inter1.get(i).end < inter2.get(j).end) {
                    total += inter1.get(i).end - inter1.get(i).start;
                    i++;
                }
                else {
                    j++;
                }
            }
        }
        while(i < n) {
            total += inter1.get(i).end - inter1.get(i).start;
            i++;
        }

        System.out.println(total - minus);

    }
}
class Interval{
    int start, end;
    Interval(int start, int end) {
        this.start = start;
        this.end = end;
    }
}